docs/ai/reference.rst

```diff
@@ -0,0 +1,357 @@
+        ...   content := 'Skies on Mars are red.'
+        ... }
+        db> insert Astronomy {
+        ...   topic := 'Earth',
+        ...   content := 'Skies on Earth are blue.'
+        ... }
+
+    ``to_context`` returns these results:
+
+    .. code-block:: edgeql-repl
+
+        db> select ext::ai::to_context(Astronomy);
+        {'Mars Skies on Mars are red.', 'Earth Skies on Earth are blue.'}
+        db> select ext::ai::to_context((select Astronomy limit 1));
+        {'Mars Skies on Mars are red.'}
+
+
+------------
+
+
+.. eql:function:: ext::ai::search( \
+                    object: anyobject, \
+                    query: <array>float32 \
+                  ) -> optional tuple<object: anyobject, distance: float64>
+
+    Search an object using its :ref:`ai::index <ref_guide_ai_reference_index>`
+    index.
+
+    Returns objects that match the specified semantic query and the
+    similarity score.
+
+    .. note::
+
+        The ``query`` argument should *not* be a textual query but the
+        embeddings generated *from* a textual query. To have EdgeDB generate
+        the query for you along with a text response, try :ref:`our built-in
+        RAG <ref_ai_overview_rag>`.
+
+    .. code-block:: edgeql-repl
+
+        db> with query := <array<float32>><json>$query
+        ...   select ext::ai::search(Knowledge, query);
+        {
+          (
+            object := default::Knowledge {id: 9af0d0e8-0880-11ef-9b6b-4335855251c4},
+            distance := 0.20410746335983276
+          ),
+          (
+            object := default::Knowledge {id: eeacf638-07f6-11ef-b9e9-57078acfce39},
+            distance := 0.7843298847773637
+          ),
+          (
+            object := default::Knowledge {id: f70863c6-07f6-11ef-b9e9-3708318e69ee},
+            distance := 0.8560434728860855
+          ),
+        }
+
+
+HTTP endpoints
+==============
+
+
+RAG
+---
+
+``https://<edgedb-host>:<port>/branch/<branch-name>/ai/rag``
+
+Responds with text generated by the specified text generation model.
+
+
+Request body
+^^^^^^^^^^^^
+
+* ``model`` (string, required): The name of the text generation model to use.
+
+* ``query`` (string, required): The query string use as the basis for text
+  generation.
+
+* ``context`` (object, required): Settings that define the context of the
+  query.
+
+  * ``query`` (string, required): Specifies the query to determine the relevant
+    objects and index to serve as context for text generation.
+
+  * ``variables`` (object, optional): A dictionary of variables for use in the
+    context query.
+
+  * ``globals`` (object, optional): A dictionary of globals for use in the
+    context query.
+
+  * ``max_object_count`` (int, optional): Maximum number of objects to return;
+    default is 5.
+
+* ``stream`` (boolean, optional): Specifies whether the response should be
+  streamed. Defaults to false.
+
+* ``prompt`` (object, optional): Settings that define a prompt, overriding the
```

Comment:
The pseudo type language here is a little bit imprecise, but maybe that's fine.

In reality this is not an object with a bunch of optional fields but rather a union of different types. Like in TypeScript this would be:

```ts
type Prompt =
  | { name: string }
  | { id: string }
  | [{ role: string; content: string }];
```

There are more precise formats like OpenAPI/Swagger or JSON Schema which we could leverage here to be perfectly precise if we don't want to adopt a specific language's type system like TypeScript.

wdyt?

Reply:
I took a look at these formats, and it seems they are intended to generate documentation, not to be user-facing documentation themselves. I guess we could try to implement this, but we would probably need to add something to the documentation build system to generate docs from one of the formats.

I do take your point here though. I missed the point of the properties of `prompt`, so I've tried to clarify that you should pass only one of the properties under prompt and which to use depending on what you want to achieve. See how you feel about the outcome.

Reply:
Yeah, I guess what we could do is not use OpenAPI or any of those tools, but see how they structure the types in the documentation that they generate. I agree that we don't have to worry too much about giving an actual machine-readable format like one of those mentioned tools, and we certainly shouldn't complicate our build process.

Maybe we could even take inspiration a bit from something like the [Stripe API docs](https://docs.stripe.com/api/balance/balance_object)? Ultimately I think we just need to add a bit more detail/precision here so that someone knows _exactly_ what they can send and _exactly_ what to expect as a response.

Reply:
Makes sense. I'll turn this over a bit and see what I can come up with.

Thank you!

Reply:
I've added some request examples, which I think could be helpful and some other minor details, like a note about authentication.

I guess the thing that's missing in terms of precision is the types of the response object properties? Is that what you're referring to when you say that it's missing detail and precision, or is there something else I'm overlooking? The request objects already give the types and descriptions of each property, so I'm not sure what else we can expose there. Aside from documentation like the Stripe ones you linked being prettier, I can't tell they're giving much more detail there. They definitely are on the responses, although I wonder how necessary it is with the examples. I think the request object details are pretty necessary since an example wouldn't capture every possible property on the object.

Reply:
In this specific case I mean that the definition for prompt is wrong in a specific kind of way. Given the current defintion you it seems like you could supply this object:

```json
{
  "prompt": {
    "name": "foo",
    "id": 123,
    "custom": [{ "role": null, "content": 42.38 }],
  }
}
```

But really this is a disjoint union like the TypeScript type I referenced earlier. So, what I'm looking for is something that makes this union-type a little more explicit. Something like "one of", or "this or that" rather than marking each property as optional. Does that make more sense? I think how Stripe would do it is to define different types of `Prompt`s and say this property is one of those types.

Reply:
Oh, the plot thickens a bit here after looking at the code. The type is actually more like:

```ts
interface CustomPrompt {
  role: "User" | "System";
  content: string;
}

type Prompt = 
  ({ name: string } | { id: string })
  & { custom: CustomPrompt[] };
```

Since you can provide a `custom` key along with the name or id (which is exclusive and will throw an error if you set both) and those messages will be appended to the prompt that you specified either by name or id.

Reply:
Ah, OK. I missed that you could append to an existing prompt. Did you see the prose I added to try to explain though? Does that resolve the other concern?

```rst
  You may specify an existing prompt by its ``name`` or ``id``, or you may
  define a custom prompt inline by sending an array of objects, each containing
  a ``role`` and ``content``. Choose to specify any one of these properties, or
  omit the ``prompt`` property to use the default prompt:

  * ``name`` (string, optional): The ``name`` of an existing custom prompt to
    use.

  * ``id`` (string, optional): The ``id`` of an existing custom prompt to use.

  * ``custom`` (array of objects, optional): Custom prompt messages, each
    containing a ``role`` and ``content``.
```

If I update that to address the ability to append to an existing prompt, does this solve the problem?

Reply:
Maybe it's not worth belaboring this, but I think my main concern is that if you squint at the definition, it's not clear that `name` and `id` are exclusive just based on looking at the structure. I think something that makes it clear structurally that `{ name: string }` and `{ id: string }` are exclusive rather than them being both optional siblings in the object's definition. Not 100% sure how to solve that (or if it's worth solving given the explanatory prose) but that's my main concern in this specific case.

But the general thing here is: how do we structure these type definitions in a friendly cross-language way that is precise enough to be non-ambiguous, but not too tied to a specific type annotation style. That's the broader question, and something we can iterate on if we're feeling stuck here.

Reply:
OK, I see. So, it's more about the implication of having the three presented as they are in the list. Yeah, that makes sense.